File: 2022/day08/main.go
package main

import (
    "bufio"
    "fmt"
    "log"
    "os"
    "strconv"
)

func handleError(err error) {
    if err != nil {
        log.Fatalf("Error: %v", err)
    }
}

func readGrid(filename string) [][]int {
    grid := make([][]int, 0)
    file, err := os.Open(filename)
    handleError(err)
    defer file.Close()

    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        line := scanner.Text()
        row := make([]int, len(line))
        for i,c := range(line) {
            row[i], _ = strconv.Atoi(string(c))
        }
        grid = append(grid, row)
    }

    handleError(scanner.Err())

    return grid
}

func isVisible(trees []int) []bool {
    visible := make([]bool, len(trees))
    maxHeight := -1
    for i,height := range(trees) {
        visible[i] = height > maxHeight
        maxHeight = max(maxHeight, height)
    }
    return visible
}

func reverse(array []int) []int {
    n := len(array)
    reversed := make([]int, n)
    for i := 0; i < len(array); i++ {
        reversed[i] = array[n-1-i]
    }
    return reversed
}

func part1(trees [][]int) int {
    visible := make([][]bool, len(trees))
    n_rows := len(trees)
    n_cols := len(trees[0])
    for i,rowTrees := range(trees) {
        visibleLeft := isVisible(rowTrees)
        visibleRight := isVisible(reverse(rowTrees))
        visible[i] = make([]bool, len(rowTrees))
        for j := 0; j < n_cols; j++ {
            visible[i][j] = visibleLeft[j] || visibleRight[n_cols-1-j]
        }
    }
    for col := 0; col < n_cols; col++ {
        colTrees := make([]int, len(trees))
        for row := 0; row < n_rows; row++ {
            colTrees[row] = trees[row][col]
        }
        visibleTop := isVisible(colTrees)
        visibleBottom := isVisible(reverse(colTrees))
        for row := 0; row < n_rows; row++ {
            visible[row][col] = visible[row][col] || visibleTop[row] || visibleBottom[n_rows-1-row]
        }
    }
    n_visible := 0
    for _,row := range(visible) {
        for _, v := range(row) {
            if v {
                n_visible++
            }
        }
    }
    return n_visible
}

func scenicScore(trees [][] int, i, j int) int {
    score := 1
    ny := len(trees)
    nx := len(trees[0])
    var x, y int
    dx := []int{0, 1, 0, -1}
    dy := []int{-1, 0, 1, 0}
    for k := 0; k < 4; k++ {
        factor := 0
        x = j + dx[k]
        y = i + dy[k]
        for {
            factor ++
            if trees[y][x] >= trees[i][j] {
                break
            }
            x += dx[k]
            y += dy[k]
            if y < 0 || y >= ny || x < 0 || x >= nx {
                break
            }
        }
        score *= factor
    }
    return score
}

func part2(trees [][]int) int {
    maxScore := 0
    for i := 1; i < len(trees)-1; i++ {
        for j := 1; j < len(trees[0])-1; j++ {
            maxScore = max(maxScore, scenicScore(trees, i, j))
        }
    }
    return maxScore
}

func main() {
    if len(os.Args) != 2 {
        fmt.Println("Usage: provide a filename")
        os.Exit(1)
    }

    filename := os.Args[1]
    grid := readGrid(filename)
    fmt.Println("Part 1:", part1(grid))
    fmt.Println("Part 2:", part2(grid))
}

File: 2022/day08/src/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define BUFFER_SIZE 256

int isVisible(int pos, int *trees, int nrows, int ncols) {
  int row = pos / ncols;
  int col = pos % ncols;
  int visibleTop = 1, visibleBottom = 1, visibleLeft = 1, visibleRight = 1;
  for (int i = 0; i < row && visibleTop; ++i) {
    visibleTop = trees[i * ncols + col] < trees[pos];
  }
  if (visibleTop)
    return 1;
  for (int i = nrows - 1; i > row && visibleBottom; --i) {
    visibleBottom = trees[i * ncols + col] < trees[pos];
  }
  if (visibleBottom)
    return 1;
  for (int i = 0; i < col && visibleLeft; ++i) {
    visibleLeft = trees[row * ncols + i] < trees[pos];
  }
  if (visibleLeft)
    return 1;
  for (int i = ncols - 1; i > col && visibleRight; --i) {
    visibleRight = trees[row * ncols + i] < trees[pos];
  }
  return visibleRight;
}

int scenicScore(int pos, int *trees, int nrows, int ncols) {
  int row = pos / ncols;
  int col = pos % ncols;
  int scoreTop = 0, scoreBottom = 0, scoreLeft = 0, scoreRight = 0;
  int i;
  for (i = row - 1; i >= 0; --i) {
    scoreTop++;
    if (trees[i * ncols + col] >= trees[pos])
      break;
  }
  for (i = row + 1; i < nrows; ++i) {
    scoreBottom++;
    if (trees[i * ncols + col] >= trees[pos])
      break;
  }
  for (i = col - 1; i >= 0; --i) {
    scoreLeft++;
    if (trees[row * ncols + i] >= trees[pos])
      break;
  }
  for (i = col + 1; i < ncols; ++i) {
    scoreRight++;
    if (trees[row * ncols + i] >= trees[pos])
      break;
  }
  return scoreTop * scoreBottom * scoreLeft * scoreRight;
}

int main(int argc, char *argv[]) {
  if (argc < 2) {
    printf("Missing input file\n");
    return 1;
  }

  FILE *file = fopen(argv[1], "r");
  if (!file) {
    printf("Error opening file\n");
    return 1;
  }
  int ncols, i, nrows = 0;
  char buffer[BUFFER_SIZE];
  int *trees;
  while (fgets(buffer, BUFFER_SIZE, file)) {
    ncols = strcspn(buffer, "\n");
    if (nrows == 0) {
      trees = (int *)malloc(ncols * 200 * sizeof(int));
    }
    for (i = 0; i < ncols; ++i) {
      trees[nrows * ncols + i] = buffer[i] - '0';
    }
    nrows++;
  }
  fclose(file);

  trees = (int *)realloc(trees, nrows * ncols * sizeof(int));
  int part1 = 0;
  int part2 = 0;
  int score;
  for (i = 0; i < nrows * ncols; ++i) {
    part1 += isVisible(i, trees, nrows, ncols);
    score = scenicScore(i, trees, nrows, ncols);
    if (score > part2) {
      part2 = score;
    }
  }
  printf("Part 1: %d\n", part1);
  printf("Part 2: %d\n", part2);

  free(trees);
  return 0;
}
